File: backend/src/DB/models/Leaf.ts
import { array, InferType, number, object, string } from 'yup';
import { likeObjectId } from '../common_schemas';

export const collectionName = 'leaf'

export const schemaVersion = 'v1.0.0'

const contentTypesSchema = string().oneOf(['string', 'richText', 'imageId', 'videoId', 'audioId'])
const contentSchema = object().shape({ type: contentTypesSchema.required(), value: array().of(string().nonUndefined().optional().nonNullable()).required().min(0) })
const contentsSchema = array().of(contentSchema.required())

const post = {
    treeNodeId: string().objectIdString().required().label('Tree node id'),
    title: string().required().label('Title'),
    termContents: contentsSchema.required().min(0).label('Term contents'),
    definitionContents: contentsSchema.required().min(0).label('Definition contents'),
}

const update = {
    _id: string().objectIdString().required().label('Id'),
    treeNodeId: string().objectIdString().optional().label('Tree node id'),
    title: string().optional().label('Title'),
    termContents: contentsSchema.optional().nonNullable().min(0).label('Term contents'),
    definitionContents: contentsSchema.optional().nonNullable().min(0).label('Definition contents'),
}

const leafCreate = {
    userId: string().objectIdString().required().label('User'),
    treeNodeId: string().objectIdString().required().label('Tree node id'),
    title: string().required().label('Title'),
    termContents: contentsSchema.required().min(0).label('Term contents'),
    definitionContents: contentsSchema.required().min(0).label('Definition contents'),
}

export const leafPostSchema = object().required().shape(post)

export const leafCreateSchema = object().required().shape(leafCreate)

export const leafUpdateSchema = object().required().shape(update)

export const leafSchema = object().required().shape(leafCreate).shape({
    schemaVersion: string().optional().min(6).max(20).label('Schema version'),
    _id: likeObjectId.optional().label('Id'),

    userId: string().objectIdString().required().label('User id'),
    treeNodeId: string().objectIdString().required().label('Tree node id'),

    title: string().required().label('Title'),
    termContents: contentsSchema.required().min(0).label('Term contents'),
    definitionContents: contentsSchema.required().min(0).label('Definition contents'),

    createdAt: number().optional().label('Created at'),
    updatedAt: number().optional().label('Updated at'),
});

export type ContentTypes = InferType<typeof contentTypesSchema>
export type Content = InferType<typeof contentSchema>
export type Contents = InferType<typeof contentsSchema>

export type LeafPost = InferType<typeof leafPostSchema>
export type LeafCreate = InferType<typeof leafCreateSchema>
export type LeafUpdate = InferType<typeof leafUpdateSchema>

export type Leaf = InferType<typeof leafSchema>

File: client/lib/api/models/leaf.dart
class LeafPost {
  final String id;

  LeafPost({required this.id});
}

enum ContentType { string, richText, imageId, videoId, audioId }

class Content {
  final ContentType type;
  final List<String> value;

  Content({required this.type, required this.value});

  static ContentType parseContentType(String value) {
    switch (value) {
      case 'string':
        return ContentType.string;
      case 'richText':
        return ContentType.richText;
      case 'imageId':
        return ContentType.imageId;
      case 'videoId':
        return ContentType.videoId;
      case 'audioId':
        return ContentType.audioId;
      default:
        throw ArgumentError('Unknown ContentType: $value');
    }
  }

  static String stringifyContentType(ContentType value) {
    switch (value) {
      case .string:
        return 'string';
      case .richText:
        return 'richText';
      case .imageId:
        return 'imageId';
      case .videoId:
        return 'videoId';
      case .audioId:
        return 'audioId';
    }
  }

  Content copyWith({ContentType? type, List<String>? value}) {
    return Content(type: type ?? this.type, value: value ?? this.value);
  }

  factory Content.fromJson(Map<String, dynamic> json) {
    final type = parseContentType(json['type'] as String);
    final value = (json['value'] as List<dynamic>).cast<String>();

    return Content(type: type, value: value);
  }

  Map<String, dynamic> toJson() {
    return {'type': stringifyContentType(type), 'value': value};
  }
}

class Leaf {
  late final String id;
  late final String userId;
  late final String treeNodeId;
  late final String title;
  late final List<Content> termContents;
  late final List<Content> definitionContents;
  late final int createdAt;
  late final int updatedAt;

  Leaf({required this.id, required this.userId, required this.treeNodeId, required this.title, required this.termContents, required this.definitionContents, required this.createdAt, required this.updatedAt});

  Leaf copyWith({String? id, String? title, List<Content>? termContents, List<Content>? definitionContents}) {
    return Leaf(id: id ?? this.id, title: title ?? this.title, termContents: termContents ?? this.termContents, definitionContents: definitionContents ?? this.definitionContents, userId: userId, treeNodeId: treeNodeId, createdAt: createdAt, updatedAt: updatedAt);
  }

  factory Leaf.fromJson(Map<String, dynamic> json) {
    final id = json['_id'];
    final userId = json['userId'];
    final treeNodeId = json['treeNodeId'];
    final title = json['title'];
    final termContents = (json['termContents'] as List<dynamic>?)?.map((e) => Content.fromJson(e as Map<String, dynamic>)).toList() ?? [];
    final definitionContents = (json['definitionContents'] as List<dynamic>?)?.map((e) => Content.fromJson(e as Map<String, dynamic>)).toList() ?? [];
    final createdAt = (json['createdAt'] as num).toInt();
    final updatedAt = (json['updatedAt'] as num).toInt();

    return Leaf(id: id, userId: userId, treeNodeId: treeNodeId, title: title, termContents: termContents, definitionContents: definitionContents, createdAt: createdAt, updatedAt: updatedAt);
  }

  Map<String, dynamic> toJson() {
    return {'id': id, 'userId': userId, 'treeNodeId': treeNodeId, 'title': title, 'termContents': termContents.map((m) => m.toJson()).toList(), 'definitionContents': definitionContents.map((m) => m.toJson()).toList(), 'createdAt': createdAt, 'updatedAt': updatedAt};
  }
}
